### ISSUE TITLE
How to save gradients during training?

### ISSUE THREAD

**Issue by louiswng** [verl-project/verl#5280]
### Feature request

I want to save the gradient at each mini-step (each update). How can I do this?

### Motivation

Analyze gradients

### Your contribution

Registered a hook, but the captured gradients are not accurate.

**Comment by Kirrito-k423** [comment]
To save gradients at each mini-step, you have a few options:

**Option 1: Register backward hooks on parameters**

```python
grad_snapshots = {}

def save_grad(name):
    def hook(grad):
        grad_snapshots[name] = grad.clone().cpu()
        return grad
    return hook

for name, param in model.named_parameters():
    if param.requires_grad:
        param.register_hook(save_grad(name))
```

**Option 2: Use forward hook on optimizer.step**

```python
from torch.optim.optimizer import Optimizer

original_step = optimizer.step

def step_with_grad_save(*args, **kwargs):
    # Gradients are still available here
    for name, param in model.named_parameters():
        if param.grad is not None:
            save_gradient(name, param.grad.clone())
    return original_step(*args, **kwargs)

optimizer.step = step_with_grad_save
```

**Option 3: For FSDP, use `state_dict` to capture sharded gradients**

If you're using FSDP with gradient sharding, the per-rank gradients are partial. Use:
```python
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
# After backward, before step
grad_dict = {name: p.grad for name, p in model.named_parameters() if p.grad is not None}
```

If your captured gradients are inaccurate, check:
1. Gradient accumulation is being accounted for
2. FSDP gradient unsharding timing (call `model.unshard_grads()` before capturing)
3. Mixed precision scaling (multiply by `scaler.get_scale()` if needed)

**Comment by louiswng** [comment]
> To save gradients at each mini-step, you have a few options:
> 
> **Option 1: Register backward hooks on parameters**
> 
> grad_snapshots = {}
> 
> def save_grad(name):
>     def hook(grad):
>         grad_snapshots[name] = grad.clone().cpu()
>         return grad
>     return hook
> 
> for name, param in model.named_parameters():
>     if param.requires_grad:
>         param.register_hook(save_grad(name))
> **Option 2: Use forward hook on optimizer.step**
> 
> from torch.optim.optimizer import Optimizer
> 
> original_step = optimizer.step
> 
> def step_with_grad_save(*args, **kwargs):
>     # Gradients are still available here
>     for name, param in model.named_parameters():
>         if param.grad is not None:
>             save_gradient(name, param.grad.clone())
>     return original_step(*args, **kwargs)
> 
> optimizer.step = step_with_grad_save
> **Option 3: For FSDP, use `state_dict` to capture sharded gradients**
> 
> If you're using FSDP with gradient sharding, the per-rank gradients are partial. Use:
> 
> from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
> # After backward, before step
> grad_dict = {name: p.grad for name, p in model.named_parameters() if p.grad is not None}
> If your captured gradients are inaccurate, check:
> 
> 1. Gradient accumulation is being accounted for
> 2. FSDP gradient unsharding timing (call `model.unshard_grads()` before capturing)
> 3. Mixed precision scaling (multiply by `scaler.get_scale()` if needed)

Thank you for your reply. I'll give it a try.

**Comment by Kirrito-k423** [comment]
中文：补充一点基于当前 `main` 代码的说明：verl 默认公开记录的是每次 optimizer update 的 `actor/grad_norm` / `critic/grad_norm`，不是逐参数梯度快照。训练调用顺序是 `optimizer_zero_grad()` -> `forward_backward_batch()` -> `optimizer_step()`；FSDP 后端会在非最后一个 micro-batch 延迟梯度同步，只在最后一个 backward 后同步，然后在 `optimizer_step()` 中 unscale/clip 并 step。因此，如果你要保存“准确的每次 update 梯度”，建议在对应 engine 的 `optimizer_step()` 中、梯度裁剪和 `optimizer.step()` 之前插桩；如果你在每个 micro-batch 的 backward hook 中保存，FSDP 下可能看到的是未同步/局部/累积中的梯度。Megatron 后端还会经过 Megatron-Core 的 `forward_backward_func` 和分布式 optimizer，逐参数全局梯度没有统一的稳定接口。

证据：`verl/workers/engine/base.py` 中 `train_batch` 的调用顺序；`verl/workers/engine/fsdp/transformer_impl.py` 中 `_gradient_sync_context` 和 `optimizer_step`；`verl/workers/engine/megatron/transformer_impl.py` 中 `forward_backward_batch` 和 `optimizer_step`；`docs/perf/perf_tuning.rst` 中关于 FSDP gradient accumulation synchronization 的说明。

English: One extra clarification based on the current `main` code: verl exposes `actor/grad_norm` / `critic/grad_norm` per optimizer update by default, not full per-parameter gradient snapshots. The training order is `optimizer_zero_grad()` -> `forward_backward_batch()` -> `optimizer_step()`. In the FSDP backend, gradient synchronization is deferred on non-final micro-batches and happens only around the final backward; then `optimizer_step()` unscales/clips gradients and performs the optimizer step. So if you need accurate per-update gradients, the safest instrumentation point is inside the relevant engine's `optimizer_step()`, before gradient clipping and before `optimizer.step()`. Capturing from backward hooks on each micro-batch can show local/unsynchronized/in-accumulation gradients under FSDP. For Megatron, the path goes through Megatron-Core `forward_backward_func` and a distributed optimizer, so there is no single backend-independent stable API for full per-parameter global gradients.

Evidence: `verl/workers/engine/base.py` for the `train_batch` order; `verl/workers/engine/fsdp/transformer_impl.py` for `_gradient_sync_context` and `optimizer_step`; `verl/workers/engine/megatron/transformer_impl.py` for `forward_backward_batch` and `optimizer_step`; `docs/perf/perf_tuning.rst` for FSDP gradient accumulation synchronization.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
